pkg/reconciler/cache/replication/replication_controller.go

```diff
@@ -0,0 +1,166 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package replication
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	kcpcache "github.com/kcp-dev/apimachinery/pkg/cache"
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	cacheclient "github.com/kcp-dev/kcp/pkg/cache/client"
+	"github.com/kcp-dev/kcp/pkg/cache/client/shard"
+	apisinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/apis/v1alpha1"
+	apislisters "github.com/kcp-dev/kcp/pkg/client/listers/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/logging"
+)
+
+const (
+	// ControllerName hold this controller name.
+	ControllerName = "kcp-replication-controller"
+
+	// AnnotationKey is the name of the annotation key used to mark an object for replication.
+	AnnotationKey = "internal.sharding.kcp.dev/replicate"
+)
+
+// NewController returns a new replication controller.
+//
+// The replication controller copies objects of defined resources that have the "internal.sharding.kcp.dev/replicate" annotation to the cache server.
+//
+// The replicated object will be placed under the same cluster as the original object.
+// In addition to that, all replicated objects will be placed under the shard taken from the shardName argument.
+// For example: shards/{shardName}/clusters/{clusterName}/apis/apis.kcp.dev/v1alpha1/apiexports
+func NewController(
+	shardName string,
+	dynamicCacheClient dynamic.ClusterInterface,
+	dynamicLocalClient dynamic.ClusterInterface,
+	localApiExportInformer apisinformers.APIExportInformer,
+	cacheApiExportInformer apisinformers.APIExportInformer,
+) (*controller, error) {
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerName)
+	c := &controller{
+		shardName:              shardName,
+		queue:                  queue,
+		dynamicCacheClient:     dynamicCacheClient,
+		dynamicLocalClient:     dynamicLocalClient,
+		localApiExportLister:   localApiExportInformer.Lister(),
+		cacheApiExportsIndexer: cacheApiExportInformer.Informer().GetIndexer(),
+	}
+
+	if err := cacheApiExportInformer.Informer().AddIndexers(cache.Indexers{
+		ByShardAndLogicalClusterAndNamespaceAndName: IndexByShardAndLogicalClusterAndNamespace,
+	}); err != nil {
+		return nil, err
+	}
+	c.cacheApiExportsIndexer = cacheApiExportInformer.Informer().GetIndexer()
+
+	localApiExportInformer.Informer().AddEventHandler(c.apiExportInformerEventHandler())
+	cacheApiExportInformer.Informer().AddEventHandler(c.apiExportInformerEventHandler())
```

Comment:
In the future, we might want to find a more efficient way of storing and comparing data from the cache. Using an informer double-down memory.